scripts/ai/AIDriveStrategyChopperCourse.lua

```diff
@@ -0,0 +1,782 @@
+            if not self:isChopperWaitingForUnloader() then
+                self:debug('I have a trailer, can continue working')
+                self:changeToFieldWork()
+            end
+        end
+        self:setMaxSpeed(0)
+    end
+end
+
+-----------------------------------------------------------------------------------------------------------------------
+--- Pipe handling
+-----------------------------------------------------------------------------------------------------------------------
+
+-- This part of an ugly workaround to make the chopper pickups work
+function AIDriveStrategyChopperCourse:checkMarkers()
+    for _, implement in pairs(AIUtil.getAllAIImplements(self.vehicle)) do
+        local aiLeftMarker, aiRightMarker, aiBackMarker = implement.object:getAIMarkers()
+        if not aiLeftMarker or not aiRightMarker or not aiBackMarker then
+            self.notAllImplementsHaveAiMarkers = true
+            return
+        end
+    end
+end
+
+-- TODO: move this to the PipeController? Rename this is it doesnt check pipe in checks for trailer in range
+function AIDriveStrategyChopperCourse:handlePipe()
+    self.pipeController:handleChopperPipe()
+end
+
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+-- Pipe offset functions
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+function AIDriveStrategyChopperCourse:setPipeOffsetX()
+    
+    
+    if self.isSugarCaneHarvester then
+        -- Since the sugarcane havester rotates we need to find total pipe length and use that as our offset. 
+        -- This math is handled in pipecontroller
+        self.pipeOffsetX = self.pipeController:getPipeOffsetX()
+    else
+        -- Use the work width plus a little bit. But make sure we don't go further than 80% of the max discarge. 
+        -- This may cause issues on very large modded choppers that didn't alter the discharge distance
+        self.pipeOffsetX =  math.min(self.chopperController:getChopperDischargeDistance() * .8, self:getWorkWidth()/2 + 4)
+    end
+    self:debug('Pipe Offset X was set as: %.2f', self.pipeOffsetX)
+end
+
+function AIDriveStrategyChopperCourse:setPipeOffsetZ(offset)
+    -- If offset is supplied use that(Chase Mode) plus what is returned from pipecontroller
+    self.pipeOffsetZ = (offset or 0) + self.pipeController:getPipeOffsetZ()
+end
+
+-- Return our current pipe offset plus any user supplied offset
+function AIDriveStrategyChopperCourse:getPipeOffset(additionalOffsetX, additionalOffsetZ)
+    self:debugSparse('Chopper PipeOffsetX is %.2f', self.pipeOffsetX)
+    return self.pipeOffsetX + additionalOffsetX, self.pipeOffsetZ + additionalOffsetZ
+end
+
+function AIDriveStrategyChopperCourse:updatePipeOffset(ix)

```

Comment:
Pipe offset for a chopper does not make sense. I understand you are trying to use that as some kind of target for the unloader, but 1) I prefer the unloader deciding where to drive and not the chopper telling it 2) but at least, it must not be called pipeOffset as it is misleading and confusing, rather unloaderTargetPosition or something like that.